src/librustc/ty/mod.rs

```diff
@@ -570,38 +572,106 @@ impl <'gcx: 'tcx, 'tcx> Canonicalize<'gcx, 'tcx> for Ty<'tcx> {
     }
 }
 
+extern {
+    /// A dummy type used to force Slice to by unsized without requiring fat pointers
+    type OpaqueSliceContents;
+}
+
 /// A wrapper for slices with the additional invariant
 /// that the slice is interned and no other slice with
 /// the same contents can exist in the same context.
 /// This means we can use pointer + length for both
 /// equality comparisons and hashing.
-#[derive(Debug, RustcEncodable)]
-pub struct Slice<T>([T]);
+pub struct Slice<T>(PhantomData<T>, OpaqueSliceContents);
```

Comment:
If you write this as:
```rust
#[repr(C)]
pub struct Slice<T> {
    pub len: usize,
    data: [T; 0],
    _opaque: OpaqueSliceContents,
}
```
Then you can read/write the `len`, and get a pointer to the start of the data by `.data.as_ptr()`, both of which work in safe code.

Reply:
+1